### PR TITLE
Adding MD to the ligand-transfer docking pipeline

### PR DESCRIPTION
**Pull request by mariacm12**
## Description
Same as https://github.com/asapdiscovery/asapdiscovery/pull/1013, but can implement MD run to the final docked position.
Current main version conflicts with the original PR so implemented as a separate one. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
- [x] Fix return_df=True issue from the older PR (csv file is not generated right now)

## Questions
- [ ] Question 1 ..

## Status
- [x] Ready to go

## Developers certificate of origin
- [ ] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

File: asapdiscovery-modeling/asapdiscovery/modeling/protein_prep.py [hunk 1]
```diff
@@ -229,12 +230,9 @@ class ProteinPrepper(ProteinPrepperBase):
...
     )
-    ref_chain: Optional[str] = Field(
-        None, description="Chain ID to align to in reference structure"
-    )
+    ref_chain: Optional[str] = Field("A", description="Reference chain ID to align to.")
     active_site_chain: Optional[str] = Field(
-        None,
-        description="Active site chain ID to align to ref_chain in reference structure",
+        "A", description="Chain ID to align to reference."
     )
```
Comment (hmacdope) [inline asapdiscovery-modeling/asapdiscovery/modeling/protein_prep.py:236]: Undo this, this has changed on `main`

File: asapdiscovery-modeling/asapdiscovery/modeling/protein_prep.py [hunk 2]
```diff
@@ -246,6 +244,21 @@ class ProteinPrepper(ProteinPrepperBase):
...
+        """
+        align = values.get("align")
+        ref_chain = values.get("ref_chain")
+        active_site_chain = values.get("active_site_chain")
+        if align and not ref_chain:
+            raise ValueError("Must provide ref_chain if align is provided")
+        if align and not active_site_chain:
+            raise ValueError("Must provide active_site_chain if align is provided")
+        return values
+
```
Comment (hmacdope) [inline asapdiscovery-modeling/asapdiscovery/modeling/protein_prep.py:261]: Also undo this, changed on `main`

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py [hunk 3]
```diff
@@ -154,8 +152,6 @@ def large_scale(
             output_dir=output_dir,
             overwrite=overwrite,
             loglevel=loglevel,
-            ref_chain=ref_chain,
-            active_site_chain=active_site_chain,
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py:158]: Unrelated changes to args? are these intentional?

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py [hunk 4]
```diff
@@ -376,11 +368,167 @@ def small_scale(
...
+    md_openmm_platform: OpenMMPlatform = OpenMMPlatform.Fastest,
+    ref_chain: Optional[str] = "A",
+    active_site_chain: Optional[str] = "A",
+    seqres_yaml: Optional[str] = None,
+    loop_db: Optional[str] = None,
+    target=target,
+):
+    """
+    Run cross docking on a set of ligands, against a set of targets.
+    """
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py:490]: Ligand transfer docking?

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py [hunk 5]
```diff
@@ -55,12 +102,19 @@ class DockingWorkflowInputsBase(BaseModel):
...
 
+    dask_cluster_n_workers: PositiveInt = Field(
+        10,
+        description="Number of workers to use as inital guess for Lilac dask cluster",
+    )
+
+    dask_cluster_max_workers: PositiveInt = Field(
+        200, description="Maximum number of workers to use for Lilac dask cluster"
+    )
+
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py:117]: These are gone now I believe?

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py [hunk 6]
```diff
@@ -75,13 +129,8 @@ class DockingWorkflowInputsBase(BaseModel):
...
     )
-    ref_chain: Optional[str] = Field(
-        None,
-        description="Chain ID to align to in reference structure containing the active site",
-    )
-    active_site_chain: Optional[str] = Field(
-        None,
-        description="Active site chain ID to align to ref_chain in reference structure",
+    walltime: str = Field(
+        "72h", description="Walltime for the workflow, used for dask-jobqueue"
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py:133]: Unrelated change?

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py [hunk 7]
```diff
@@ -121,18 +170,6 @@ def check_inputs(cls, values):
...
-        active_site_chain = values.get("active_site_chain")
-        ref_chain = values.get("ref_chain")
-        target = values.get("target")
-        if not active_site_chain:
-            values["active_site_chain"] = active_site_chains[target]
-        # set same chain for active site if not specified
-        if not ref_chain:
-            values["ref_chain"] = active_site_chains[target]
-        return values
-
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py:135]: Unrelated change

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py [hunk 8]
```diff
@@ -6,13 +6,60 @@
...
+    class Config:
+        arbitrary_types_allowed = True
+
+    @classmethod
+    def from_json_file(cls, file: str | Path):
+        return cls.parse_file(str(file))
+
+    def to_json_file(self, file: str | Path):
+        with open(file, "w") as f:
+            f.write(self.json(indent=2))
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/workflows.py:59]: Is it necessary to change this / was this intentional?

File: asapdiscovery-modeling/asapdiscovery/modeling/protein_prep.py [hunk 9]
```diff
@@ -268,8 +281,6 @@ def _prep(self, inputs: list[Complex], failure_mode="skip") -> list[PreppedCompl
                 if self.seqres_yaml:
                     with open(self.seqres_yaml) as f:
                         seqres_dict = yaml.safe_load(f)
-                    if "SEQRES" not in seqres_dict:
-                        raise ValueError("No SEQRES found in YAML")
```
Comment (hmacdope) [inline asapdiscovery-modeling/asapdiscovery/modeling/protein_prep.py:272]: Same here.

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+        dask_client=dask_client,
+        return_df=False,
+        failure_mode=inputs.failure_mode,
+    )
+    import csv
+
+    with open(data_intermediates / "docking_scores_raw.csv", "w", newline="") as myfile:
+        wr = csv.writer(myfile, quoting=csv.QUOTE_ALL)
+        for item in scores_list:
+            wr.writerow([item])
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py:369]: Write with `pandas` not `csv` package.

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+        logger.info("Writing final docked poses to SDF file")
+        write_ligands_to_multi_sdf(
+            output_dir / "docking_results.sdf", [r.posed_ligand for r in results]
+        )
+    # NOTE: The return_df option doesn't work, the DataFrame.pivot() function fails. Fix is pending.
+    scores_list = scorer.score(
+        results,
+        use_dask=inputs.use_dask,
+        dask_client=dask_client,
+        return_df=False,
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py:361]: If you return a `pandas.Dataframe` here everything will be easier (I promise!).

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+            simulation_results,
+            use_dask=inputs.use_dask,
+            dask_client=dask_client,
+            failure_mode=inputs.failure_mode,
+        )
+        gifs.to_csv(data_intermediates / "md_gifs.csv", index=False)
+        # duplicate target id column so we can join
+        gifs[DockingResultCols.DOCKING_STRUCTURE_POSIT.value] = gifs[
+            DockingResultCols.TARGET_ID.value
+        ]
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py:446]: Looking good overall.

**Review by hmacdope — CHANGES_REQUESTED** [review]
I think there are a few unrelated changes mixed in possibly due to basing this not on latest `main` but the main functionality looks sound. You may want to try and merge with upstream / main or rebase PR.

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py
Comment (mariacm12) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py:1]: Copied this from the latest version of the code from `main`, so not sure why it says I made changes to the cli?

Comment (hmacdope) [reply asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/cli.py:1]: Did you copy from `main` or `upstream/main`. My guess is that you are behind. Regardless, the best way to sort this out is to merge with upstream/main rather than copying over files. e.g `git fetch upstream && git merge upstream/main`

**Comment by mariacm12** [comment]
@hmacdope This should be good to go. The code is updated with the latest version and I fixed the problem with the `return_df` option (the pipeline generates all the outputs now)

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+
+    if inputs.save_to_cache and inputs.cache_dir is not None:
+        logger.info(f"Writing prepped complexes to global cache {inputs.cache_dir}")
+        prepper.cache(prepped_complexes, inputs.cache_dir)
+
+    # Here the only thing that makes sense is the self docking selector
+    selector = StructureSelector.SELF_DOCKING.selector_cls()
+
+    # If the same reference is used for multiple targets, we will have duplicated ligands. This give problems when calling df.pivot()
+    def get_unique_ligands(complex_list):
```
Comment (mariacm12) [inline asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py:298]: Had to add a function to check for repeated ligands, because the way the proteins are prepped right now is with a loop over the target x reference complexes, to assign each ligand from the reference to a target. So if we have multiple targets, we end up having a ligand repeated `n_targets` times, which gives problem when calling df.pivot() due to repeated entries. 
Not sure where is the best place to put this function, should I live it there?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looks good overall, but I think I need a talk through what the issues is with duplicated ligands. Also needs tests.

File: asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py (hunk 13 above)
Comment (hmacdope) [reply asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py:298]: Yeah I am happy for it to live here, close to the call site makes it easy to understand.

**Comment by hmacdope** [comment]
> Looks good overall, but I think I need a talk through what the issues is with duplicated ligands. Also needs tests.

I understand this now, should be good to go ahead with some tests.

**Comment by hmacdope** [comment]
@mariacm12 to detail this further what is happening in the absence of the `unique_ligands` function (you probably have this figured out but just for the record).

Example scenario:

- We have 5 structures and one reference complex with a ligand
- Once prepped the 5 complexes all have the same reference ligand 
- For each of the prepped complexes we extract its ligand (5 of the same ligand in this case).
- For each ligand, prepped complexes with a matching ligand are selected by the SelfDockingSelector (all 5 match so we have 5x5 = 25 pairs). 

This means your approach of selecting unique ligands is the correct one AFAIK. This does not need changes to selectors. 

We can do it slightly more robustly by using inchikeys but other than that all good. 

I have pushed some small changes but otherwise looks good, just needs tests.

File: asapdiscovery-workflows/asapdiscovery/workflows/tests/conftest.py [hunk 14]
```diff
@@ -95,3 +95,8 @@ def mac1_complex():
     return PreppedComplex.parse_file(
         fetch_test_file("constrained_conformer/complex.json")
     )
+
+
+@pytest.fixture()
+def alphafold_maestro_prep_pdb():
+    return fetch_test_file("Mpro-YP_009725301_AFold_processed.pdb")
```
Comment (hmacdope) [inline asapdiscovery-workflows/asapdiscovery/workflows/tests/conftest.py]: @mariacm12 you should be able to use the file from this fixture.

File: asapdiscovery-workflows/asapdiscovery/workflows/tests/docking_cli/conftest.py
Comment (mariacm12) [inline asapdiscovery-workflows/asapdiscovery/workflows/tests/docking_cli/conftest.py:1]: It seems like all the docking workflow tests refer to fixtures in `asapdiscovery-workflows/asapdiscovery/workflows/tests/docking_cli/conftest.py`, rather than `asapdiscovery-workflows/asapdiscovery/workflows/tests/conftest.py`. Thus, I included the fixture in the first file. Can you confirm that's ok or should I leave it as you suggested?

Comment (hmacdope) [reply asapdiscovery-workflows/asapdiscovery/workflows/tests/docking_cli/conftest.py:1]: Either is fine, conftest inherits from conftests in the directories above them.

**Comment by hmacdope** [comment]
To make record of a decision made here. We are hitting some kind of runner limit and there isn't much we can do about it. @mariacm12 and I decided not to block on this and add a skip for this test, we will have to rely on usage to battle test.

**Review by hmacdope — APPROVED** [review]
